### PR TITLE
Modify service setup URL method to rely on its v2 version.

### PR DESCRIPTION
**Pull request by techanvil**
## Summary

Addresses issue #4376 

## Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [ ] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @techanvil. This almost looks good to me. Just a few nitpicks and this is ready to go.

**Comment by techanvil** [comment]
Thanks for the review @eugene-manuilov, I've made the changes as requested :+1:

File: includes/Core/Authentication/Authentication.php [hunk 1]
```diff
@@ -1087,12 +1088,27 @@ private function get_authentication_oauth_error_notice() {
...
-					} elseif ( $this->credentials->using_proxy() ) {
-						$access_code = $this->user_options->get( OAuth_Client::OPTION_PROXY_ACCESS_CODE );
-						$setup_url = $auth_client->get_proxy_setup_url( $access_code );
+					$access_code = $this->user_options->get( OAuth_Client::OPTION_PROXY_ACCESS_CODE );
+
+					if ( Feature_Flags::enabled( 'serviceSetupV2' ) ) {
+						$is_using_proxy = $this->credentials->using_proxy() && $access_code;
+					} else {
+						$is_using_proxy = $this->credentials->using_proxy();
+					}
```
[the remark this answers is not in the document]
Comment (techanvil) [reply includes/Core/Authentication/Authentication.php]: No problem, I've made this change, cheers.

File: includes/Core/Authentication/Authentication.php [hunk 2]
```diff
@@ -1087,12 +1088,27 @@ private function get_authentication_oauth_error_notice() {
...
+
+					if ( $is_using_proxy ) {
+						if ( Feature_Flags::enabled( 'serviceSetupV2' ) ) {
+							$params = array(
+								'code'    => $access_code,
+								'site_id' => $this->credentials->get()['oauth2_client_id'],
```
[the remark this answers is not in the document]
Comment (techanvil) [reply includes/Core/Authentication/Authentication.php]: This is done too :+1:

File: includes/Core/Authentication/Google_Proxy.php [hunk 3]
```diff
@@ -148,6 +150,53 @@ public function setup_url( Credentials $credentials, array $query_params = array
 		return add_query_arg( $params, $this->url( self::SETUP_URI ) );
 	}
 
+	/**
+	 * Returns the setup URL to the authentication proxy.
+	 *
+	 * This should be renamed to replace `setup_url` once the `serviceSetupV2` feature is fully developed and the feature flag is removed.
```
[the remark this answers is not in the document]
Comment (techanvil) [reply includes/Core/Authentication/Google_Proxy.php]: Done :+1:

File: includes/Core/Authentication/Setup_V1.php [hunk 4]
```diff
@@ -89,18 +90,26 @@ public function handle_action_setup() {
...
 		if ( $verification_token && $verification_method ) {
 			$this->handle_verification( $verification_token, $verification_method );
+
+			$proxy_query_params = array(
+				'verify'              => 'true',
+				'verification_method' => $verification_method,
+			);
+		} else {
+			$proxy_query_params = array();
 		}
```
[the remark this answers is not in the document]
Comment (techanvil) [reply includes/Core/Authentication/Setup_V1.php]: Done :+1:

File: includes/Core/Authentication/Google_Proxy.php [hunk 5]
```diff
@@ -148,6 +150,53 @@ public function setup_url( Credentials $credentials, array $query_params = array
...
+	 * @since n.e.x.t
+	 *
+	 * @param array $query_params Query parameters to include in the URL.
+	 * @return string URL to the setup page on the authentication proxy.
+	 *
+	 * @throws Exception Thrown if called without the required query parameters.
+	 */
+	public function setup_url_v2( array $query_params = array() ) {
+		if ( empty( $query_params['code'] ) ) {
+			throw new Exception( __( 'Missing code parameter for setup URL.', 'google-site-kit' ) );
```
Comment (eugene-manuilov) [inline includes/Core/Authentication/Google_Proxy.php:167]: When I start the setup process on my local machine, I get the following fatal error:

```
PHP Fatal error:  Uncaught Exception: Missing code parameter for setup URL. in /var/www/html/wp-content/plugins/google-site-kit/includes/Core/Authentication/Google_Proxy.php:167
Stack trace:
#0 /var/www/html/wp-content/plugins/google-site-kit/includes/Core/Authentication/Setup.php(189): Google\Site_Kit\Core\Authentication\Google_Proxy->setup_url_v2(Array)
#1 /var/www/html/wp-content/plugins/google-site-kit/includes/Core/Authentication/Setup_V2.php(65): Google\Site_Kit\Core\Authentication\Setup->redirect_to_proxy()
#2 /var/www/html/wp-includes/class-wp-hook.php(303): Google\Site_Kit\Core\Authentication\Setup_V2->handle_action_setup_start('')
#3 /var/www/html/wp-includes/class-wp-hook.php(327): WP_Hook->apply_filters('', Array)
#4 /var/www/html/wp-includes/plugin.php(470): WP_Hook->do_action(Array)
#5 /var/www/html/wp-admin/admin.php(408): do_action('admin_action_go...')
#6 /var/www/html/wp-admin/index.php(10): require_once('/var/www/html/w...')
#7 {main}
  thrown in /var/www/html/wp-content/plugins/google-site-kit/includes/Core/Authentication/Google_Proxy.php on line 167
```

Does it work for you? Can you set up the plugin with the `serviceSetupV2` feature flag enabled?

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @techanvil. The code looks good to me, but doesn't work when I try to connect the plugin on my local machine. Can you take a look?

File: includes/Core/Authentication/Google_Proxy.php (hunk 5 above)
Comment (techanvil) [reply includes/Core/Authentication/Google_Proxy.php:167]: Unfortunately, like its predecessor #4383, the V2 aspect of this ticket is not directly testable, as the flow relies on work still to come in #2044 (and possibly #2045 - note how the AC for #2045 mentions _"While the changes in the ACs here are minimal, this issue should be used for a holistic QA check of the entire feature"_).

Comment (eugene-manuilov) [reply includes/Core/Authentication/Google_Proxy.php:167]: @techanvil this particular exception is thrown in the `Setup_V2::handle_action_setup_start` method that initiates the setup process and redirects the user to the proxy server. The missing `code` parameter is returned by the proxy, thus there is no way to get it before going to the proxy server. So, this is a legit bug that we need to fix. Users should be able to redirect to the proxy without having the `code` parameter at the beginning. cc @felixarntz

Comment (techanvil) [reply includes/Core/Authentication/Google_Proxy.php:167]: Hi @eugene-manuilov - my understanding is that this existing code within `Setup_V2::handle_action_setup_start`, which leads to the exception being thrown:

https://github.com/google/site-kit-wp/blob/0305b070dacf5fb4091091499ad3bf8cce6c77ed/includes/Core/Authentication/Setup_V2.php#L57-L65

Will be replaced with code along the lines of what's in the PoC here:

https://github.com/google/site-kit-wp/blob/0e6f380313516412df5435bb255cb6083551512f/includes/Core/Authentication/Authentication.php#L1338-L1350

I.e. redirecting to the URL returned by either the `sync_site_fields` or `register_site` method - both of which are due to be implemented in #2044 (`sync_site_fields` to be updated and `register_site` to be implemented in full).

It's a bit unfortunate that we're left with things as they are for now, but I don't see how we can address this bug without updating `sync_site_fields` / implementing `register_site` which lie outside the scope of this ticket...

I may have things wrong here, but that's as far as my understanding of the situation goes, happy to be corrected if I am mistaken. It's possible we might need to get @felixarntz's input on this?

Comment (eugene-manuilov) [reply includes/Core/Authentication/Google_Proxy.php:167]: Ok, let's have this sorted out in #2044 then. 👍

**Review by eugene-manuilov — APPROVED** [review]

**Pull request merged**